Const correctness and other problems in GetGameName

src/doom/d_main.c:849:43: error: passing argument 1 of ‘GetGameName’ discards ‘const’ qualifier from pointer target type [-Werror=discarded-qualifiers]

```
             gamedescription = GetGameName("Freedoom: Phase 1");
```

`GetGameName` is here:

https://github.com/chocolate-doom/chocolate-doom/blob/493194672320732fd03e28f489672d76f4a9d58d/src/doom/d_main.c#L676

I can't just make `gamename` `const` since it's also used for building the dehacked description.

Since the function sometimes returns an allocated string and sometimes the original constant the caller can't free it if necessary. This causes a memory leak, although a bounded one.

What's the best way to fix these issues?

The string returned from `DEH_String` is used as a format string without verification. This can cause problems, even security issues. Do we care about malicious DEHs?

Also the return value of `GetGameName` is stored in `gamedescription` which is declared `extern` in `doomstat.h` and defined in `doomstat.c`. However it doesn't appear to be used outside `d_main.c`. Heretic and Hexen have it in their respective `*_main.c` and declared `static`. Can I move it and remove the `extern` definition?'

This code has been copy-pasted into Strife so it should be fixed there too.

--- a/src/doom/d_main.c
+++ b/src/doom/d_main.c
@@ -673,7 +673,7 @@ static const char *banners[] =
 // Otherwise, use the name given
 // 
 
-static char *GetGameName(char *gamename)
+static char *GetGameName(const char *gamename)
 {
     size_t i;
     const char *deh_sub;

--- a/src/doom/d_main.c
+++ b/src/doom/d_main.c
@@ -718,7 +718,7 @@ static char *GetGameName(char *gamename)
         }
     }
 
-    return gamename;
+    return M_StringDuplicate(gamename);
 }
 
 static void SetMissionForPackName(const char *pack_name)

--- a/src/doom/d_main.c
+++ b/src/doom/d_main.c
@@ -678,14 +678,14 @@ static const char *banners[] =
 static char *GetGameName(const char *gamename)
 {
     size_t i;
-    
+
     for (i=0; i<arrlen(banners); ++i)
     {
         const char *deh_sub;
         // Has the banner been replaced?
 
         deh_sub = DEH_String(banners[i]);
-        
+
         if (deh_sub != banners[i])
         {
             size_t gamename_size;

--- a/src/doom/d_main.c
+++ b/src/doom/d_main.c
@@ -676,10 +676,10 @@ static const char *banners[] =
 static char *GetGameName(const char *gamename)
 {
     size_t i;
-    const char *deh_sub;
     
     for (i=0; i<arrlen(banners); ++i)
     {
+        const char *deh_sub;
         // Has the banner been replaced?
 
         deh_sub = DEH_String(banners[i]);

--- a/src/doom/d_main.c
+++ b/src/doom/d_main.c
@@ -87,7 +87,7 @@
 //
 void D_DoomLoop (void);
 
-static const char *gamedescription;
+static char *gamedescription;
 
 // Location where savegames are stored
 

--- a/src/doom/d_main.c
+++ b/src/doom/d_main.c
@@ -87,6 +87,8 @@
 //
 void D_DoomLoop (void);
 
+static const char *gamedescription;
+
 // Location where savegames are stored
 
 char *          savegamedir;

--- a/src/doom/d_main.c
+++ b/src/doom/d_main.c
@@ -688,28 +688,29 @@ static char *GetGameName(char *gamename)
         {
             size_t gamename_size;
             int version;
+            char *deh_gamename;
 
             // Has been replaced.
             // We need to expand via printf to include the Doom version number
             // We also need to cut off spaces to get the basic name
 
             gamename_size = strlen(deh_sub) + 10;
-            gamename = Z_Malloc(gamename_size, PU_STATIC, 0);
+            deh_gamename = Z_Malloc(gamename_size, PU_STATIC, 0);
             version = G_VanillaVersionCode();
-            M_snprintf(gamename, gamename_size, deh_sub,
+            M_snprintf(deh_gamename, gamename_size, deh_sub,
                        version / 100, version % 100);
 
-            while (gamename[0] != '\0' && isspace(gamename[0]))
+            while (deh_gamename[0] != '\0' && isspace(deh_gamename[0]))
             {
-                memmove(gamename, gamename + 1, gamename_size - 1);
+                memmove(deh_gamename, deh_gamename + 1, gamename_size - 1);
             }
 
-            while (gamename[0] != '\0' && isspace(gamename[strlen(gamename)-1]))
+            while (deh_gamename[0] != '\0' && isspace(deh_gamename[strlen(deh_gamename)-1]))
             {
-                gamename[strlen(gamename) - 1] = '\0';
+                deh_gamename[strlen(deh_gamename) - 1] = '\0';
             }
 
-            return gamename;
+            return deh_gamename;
         }
     }
 

--- a/src/doom/d_main.c
+++ b/src/doom/d_main.c
@@ -845,8 +845,6 @@ void D_IdentifyVersion(void)
 
 static void D_SetGameDescription(void)
 {
-    gamedescription = "Unknown";
-
     if (logical_gamemission == doom)
     {
         // Doom 1.  But which version?
@@ -894,6 +892,11 @@ static void D_SetGameDescription(void)
         {
             gamedescription = GetGameName("DOOM 2: TNT - Evilution");
         }
+    }
+
+    if (gamedescription == NULL)
+    {
+        gamedescription = M_StringDuplicate("Unknown");
     }
 }
 

--- a/src/doom/d_main.c
+++ b/src/doom/d_main.c
@@ -695,7 +695,11 @@ static char *GetGameName(char *gamename)
             // We also need to cut off spaces to get the basic name
 
             gamename_size = strlen(deh_sub) + 10;
-            deh_gamename = Z_Malloc(gamename_size, PU_STATIC, 0);
+            deh_gamename = malloc(gamename_size);
+            if (deh_gamename == NULL)
+            {
+                I_Error("GetGameName: Failed to allocate new string");
+            }
             version = G_VanillaVersionCode();
             M_snprintf(deh_gamename, gamename_size, deh_sub,
                        version / 100, version % 100);

--- a/src/doom/d_main.c
+++ b/src/doom/d_main.c
@@ -701,8 +701,8 @@ static char *GetGameName(const char *gamename)
                 I_Error("GetGameName: Failed to allocate new string");
             }
             version = G_VanillaVersionCode();
-            M_snprintf(deh_gamename, gamename_size, deh_sub,
-                       version / 100, version % 100);
+            DEH_snprintf(deh_gamename, gamename_size, banners[i],
+                         version / 100, version % 100);
 
             while (deh_gamename[0] != '\0' && isspace(deh_gamename[0]))
             {

--- a/src/doom/d_main.c
+++ b/src/doom/d_main.c
@@ -843,7 +843,7 @@ void D_IdentifyVersion(void)
 
 // Set the gamedescription string
 
-void D_SetGameDescription(void)
+static void D_SetGameDescription(void)
 {
     gamedescription = "Unknown";
 

--- a/src/doom/doomstat.c
+++ b/src/doom/doomstat.c
@@ -26,7 +26,6 @@ GameMode_t gamemode = indetermined;
 GameMission_t	gamemission = doom;
 GameVersion_t   gameversion = exe_final2;
 GameVariant_t   gamevariant = vanilla;
-const char *gamedescription;
 
 // Set if homebrew PWAD stuff has been added.
 boolean	modifiedgame;

--- a/src/doom/doomstat.h
+++ b/src/doom/doomstat.h
@@ -57,7 +57,6 @@ extern GameMode_t	gamemode;
 extern GameMission_t	gamemission;
 extern GameVersion_t    gameversion;
 extern GameVariant_t    gamevariant;
-extern const char       *gamedescription;
 
 // Convenience macro.
 // 'gamemission' can be equal to pack_chex or pack_hacx, but these are

